core/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -232,6 +233,16 @@ public CompiledScript compile(Script script, ScriptContext scriptContext) {
         if (canExecuteScript(lang, scriptEngineService, script.getType(), scriptContext) == false) {
             throw new ScriptException("scripts of type [" + script.getType() + "], operation [" + scriptContext.getKey() + "] and lang [" + lang + "] are disabled");
         }
+
+        // special exception to prevent expressions from compiling as update or mapping scripts
+        boolean expression = scriptEngineService instanceof ExpressionScriptEngineService;
```

Comment:
I'm in complete agreement with this and had thought about doing something along these lines.  It will require both a breaking change and a great deal of work to accomplish at this time.  We have to figure out how global settings will affect things that aren't allowed at all.  Also, many tests rely on the fact that all scripts can possibly run in all modes, so those will have to be updated as well.  Given the goal for this PR is to be released as quickly as possible for the beta, I'm going to push this out to a different PR.